# MỤC LỤC
**Câu 1 (2 điểm)**
* a/ Lập luận lý thuyết & Mô phỏng.
* b/ Lập luận lý thuyết & Mô phỏng.

**Câu 3 (3 điểm)**
* a/ Lập luận lý thuyết & Mô phỏng.
* b/ Lập luận lý thuyết & Mô phỏng.
* c/ Lập luận lý thuyết & Mô phỏng.
* d/ Mô phỏng.

**Câu 4 (5 điểm)**
* a/ Lập luận lý thuyết & Mô phỏng.
* b/ Lập luận lý thuyết & Mô phỏng.
* c/ Mô phỏng.
* d/ Lập luận lý thuyết & Mô phỏng.
* e/ Lập luận lý thuyết & Mô phỏng.
* f/ Lập luận lý thuyết & Mô phỏng.
* g/ Lập luận lý thuyết & Mô phỏng.
* h/ Lập luận lý thuyết & Mô phỏng.


# Câu 1 

In [1]:
# import thư viện cần thiết
import numpy as np
import random as rd

In [2]:
def randpercent(): # random phần trăm
    rand = rd.uniform(0, 1)
    return rand

**a/ Biết router R1 bị hỏng, tính xác suất A, B liên lạc được.**

**TH: R2, R3 hoạt động, R4 hỏng:**
* Xác suất: (1-0.15)*(1-0.2)*0.25 = 0.17        (1)

**TH: R2, R4 hoạt động, R3 hỏng:**
* Xác suất: (1-0.15)*(1-0.25)*0.2 = 0.1275      (2)

**TH: R2, R3, R4 hoạt động:**
* Xác suất: (1-0.15)*(1-0.2) *(1-0.25) = 0.51   (3)

**R1 hỏng, xác suất A B liên lạc được là:** 
* (1) + (2) + (3) = 0.17 + 0.1275 + 0.51 = 0.8075

In [3]:
# Mô phỏng
def check_connection_a(n):
    cc = 0
    for _ in range(n):
        R = np.array([0, 0, 0, 0, 0])
        for i in range(1, 5):
            percent = randpercent()
            if percent > pr_fail[i]:
                R[i] = 1
        if ((R[2] == R[3] == 1) or (R[2] == R[4] == 1)):
            cc += 1

    return cc

pr_fail = np.array([0, 1, 0.15, 0.20, 0.25]) # cho xác suất hỏng cùa R[1] = 1
n = 100000
print(check_connection_a(n)/n)

0.80764


**b/ Tính xác suất A, B liên lạc được**

**TH R1 hỏng, kế thừa kết quả câu a:**
* Xác suất: 0.1*0.8075 = 0.08075    (1)

**TH R1 hoạt động, chắc chắn liên lạc được:**
* Xác suất: 1-0.1 = 0.9             (2)

**Xác suất A, B liên lạc được là:**
* (1) + (2) = 0.08075 + 0.9 = 0.98075

In [4]:
# Mô phỏng

def check_connection_b(n):
    cc = 0
    for _ in range(n):
        R = np.array([0, 0, 0, 0, 0])
        for i in range(1, 5):
            percent = randpercent()
            if percent > pr_fail[i]:
                R[i] = 1
        if ((R[2] == R[3] == 1) or (R[2] == R[4] == 1) or (R[1] == 1)):
            cc += 1

    return cc

pr_fail = np.array([0, 0.10, 0.15, 0.20, 0.25]) # Xác suất bị hỏng các R
n = 100000
check_connection_b(n)/n

0.98103

# Câu 2

In [5]:
# import thư viện cần thiết
import numpy as np
import random as rd

In [6]:
def randpercent():  # random phần trăm
    rand = rd.uniform(0, 1)
    return rand

In [7]:
def plus_one(x): # Cộng thêm 1 cho x
    x += 1
    return x

**a/ Giả sử con chuột đang ở Phòng 1, tính xác suất con chuột vẫn ở Phòng 1 sau đó 5 ngày.**

* Ta lập ma trận xác suất chuyển phòng mỗi ngày của chuột
$$ P = 
\begin{pmatrix}
1/2 & 1/4 & 0 & 0 \\
1/2 & 1/4 & 1/3 & 1/3 \\
0 & 1/4 & 1/3 & 1/3 \\
0 & 1/4 & 1/3 & 1/3
\end{pmatrix} $$
* Gọi Xt là xích markov chỉ phòng chuột ở ngày t.
* ycbt <=> $P$($X_5$ = 1| $X_0$ = 1) = $(P^{5})_{1,1}$
* 
$$P^{5} = 
\begin{pmatrix}
0.20334201 & 0.1714048 & 0.15128279 & 0.15128279 \\
0.34280961 & 0.3345028 & 0.32939493 & 0.32939493 \\
0.22692419 & 0.2470462 & 0.25966114 & 0.25966114 \\
0.22692419 & 0.2470462 & 0.25966114 & 0.25966114
\end{pmatrix}
$$
* $(P^{5})_{1,1}$ = 0.20334201

In [8]:
# Tính P^5 cho cách giải lý thuyết
def matmultiply(a, b):
    res = np.zeros(a.shape)
    size = a.shape[0]
    for i in range(size):
        for j in range(size):
            for k in range(size):
                res[i][j] += a[i][k] * b[k][j]

    return res

p = np.array([[1/2, 1/4, 0, 0],
            [1/2, 1/4, 1/3, 1/3],
            [0, 1/4, 1/3, 1/3],
            [0, 1/4, 1/3, 1/3]] )

p_2 = matmultiply(p, p)
p_3 = matmultiply(p_2,p)
p_5 = matmultiply(p_3,p_2)
print(p_5)

[[0.20334201 0.1714048  0.15128279 0.15128279]
 [0.34280961 0.3345028  0.32939493 0.32939493]
 [0.22692419 0.2470462  0.25966114 0.25966114]
 [0.22692419 0.2470462  0.25966114 0.25966114]]


In [9]:
# Mô phỏng
quantity = [0, 2, 4, 3, 3] # lưu số lượng lựa chọn di chuyển phòng thứ i
index = np.array([[0, 0, 0, 0, 0],
                [0, 1, 2, 0, 0],
                [0, 1, 2, 3, 4],
                [0, 2, 3, 4, 0],
                [0, 2, 3, 4, 0]]) # Ma trận lưu các phòng có thể duy chuyển tại mỗi phòng

sl = 0
n = 100000 # lần lặp
for _ in range(n):
    start = 1
    for i in range(5):
        percent = randpercent()
        for j in range(1, quantity[start]+1):
            if (percent <= j*(1/quantity[start])):
                start = index[start][j]
                break
    if start == 1:
        sl += 1

print(sl/n)

0.20222


**b/ Sau rất nhiều ngày, xác suất con chuột ở trong mỗi phòng là bao nhiêu?**

* Gọi vector 
$\pi = 
\begin{pmatrix}
a \\
b \\
c \\
d
\end{pmatrix}
$ là xác xuất chuột ở mỗi phòng sau nhiều ngày
* ycbt <=> $P\pi = \pi$ <=> ($P$ - $I_5$)$\pi$ = 0
$$ 
P - I_5 = 
\begin{pmatrix}
-1/2 & 1/4 & 0 & 0 \\
1/2 & -3/4 & 1/3 & 1/3 \\
0 & 1/4 & -2/3 & 1/3 \\
0 & 1/4 & 1/3 & -2/3
\end{pmatrix} $$
$\longrightarrow$
$\begin{pmatrix}
-1/2 & 1/4 & 0 & 0 \\
1/2 & -3/4 & 1/3 & 1/3 \\
0 & 1/4 & -2/3 & 1/3 \\
0 & 1/4 & 1/3 & -2/3
\end{pmatrix}$

$d_1 + d_2, 2d_3 + d_2, 2d_4 + d_2 + d_3\longrightarrow$
$\begin{pmatrix}
-1/2 & 1/4 & 0 & 0 \\
0 & -1/2 & 1/3 & 1/3 \\
0 & 0 & -1 & 1 \\
0 & 0 & 0 & 0
\end{pmatrix}$
$\longrightarrow$
$
\begin{cases}
a &= 2\alpha/3 \\
b &= 4\alpha/3 \\
c &= \alpha \\
d &= \alpha
\end{cases}
$
* Mà a + b + c + d = 1 -> $\alpha$ = 1/4

$\longrightarrow$
$
\begin{cases}
a &= 1/6 \\
b &= 1/3 \\
c &= 1/4 \\
d &= 1/4
\end{cases}
$
* Vậy xác xuất chuột ở phòng 1, 2, 3, 4 lần lượt là: 1/6, 1/3, 1/4, 1/4

In [10]:
# Mô phỏng
def plus_one(x):
    x += 1
    return x

quantity = [0, 2, 4, 3, 3] #
index = np.array([[0, 0, 0, 0, 0],
                [0, 1, 2, 0, 0],
                [0, 1, 2, 3, 4],
                [0, 2, 3, 4, 0],
                [0, 2, 3, 4, 0]] )#

sl = np.array([0, 0, 0, 0, 0]) # Mảng để lưu số lượng chuột kết thúc ở phòng thứ i 
n = 10000 # số lần lặp 

for _ in range(n):
    start = rd.randint(1, 4) # random phòng thả chuột
    day = rd.randint(200, 300) # random số ngày 
    for i in range(day):
        percent = randpercent()
        for j in range(1, quantity[start]+1):
            if (percent <= j*(1/quantity[start])):
                start = index[start][j]
                break
    sl[start] = plus_one(sl[start])

print(sl[1]/n, sl[2]/n, sl[3]/n, sl[4]/n)

0.165 0.3365 0.2456 0.2529


**c/ Giả sử con chuột ở mỗi phòng không quá 3 ngày liên tiếp (nghĩa là nếu con chuột ở một phòng nào đó đã 3 ngày liên tiếp thì ngày hôm sau sẽ chuyển đến phòng bên với xác suất đều mà không ở lại phòng đó nữa), làm lại Câu (a).**

**Dùng QHĐ tại ngày n của mỗi phòng kế thừa các phòng ngày (n-1) có thể đi đến nó ta được bảng xác suất:**
|  | **phòng 1** | **phòng 2** |**phòng 3**|**phòng 4**|
| --- | --- | --- | --- | --- |
| **Ngày 1** | 1/2 | 1/2 | 0 | 0 |
| **Ngày 2** | 3/8 | 3/8 | 1/8 | 1/8 |
| **Ngày 3** |  |  | 17/96 | 17/96 |

**Theo đề chuột không ở một phòng quá 3 ngày.**
* Ngày 3 phòng 1 đã có TH chuột ở phòng 1 được 3 ngày trước đó nên có xác suất là: $(3/8 - {(1/2)}^2)* 1/2 + 3/8* 1/4 = 5/32 $ (Ta kế thừa đi đến phòng 1 ngày 3 từ phòng 1, 2 ngày 2 có trừ xs ở phòng 1 3 ngày liền)
* Ngày 3 phòng 2 có xác suất là: $1 - 5/32 - (17/96)* 2$ = 47/96 (Vì tổng xs các phòng trong 1 ngày là 1)

|  | **phòng 1** | **phòng 2** |**phòng 3**|**phòng 4**|
| --- | --- | --- | --- | --- |
| **Ngày 3** | 5/32 | 47/96 | 17/96 | 17/96 |
* Ngày 4 phòng 1 có xác suất là: $(5/32)* 1/2 + (47/96 - 1/2* {(1/4)}^2)* 1/4 + (1/2* {(1/4)}^2)* 1/3 = 13/64 $ (TH chuột ở phòng 2 3 ngày liền thì xs qua phòng 1 là 1/3, TH không ở liền 3 ngày thì xs nhân 1/4)
* Ngày 4 phòng 3, 4 có cùng xác suất là: $ 17/96* 2/3 + (47/96 - 1/2* {(1/4)}^2)* 1/4 + (1/2* {(1/4)}^2)* 1/3 = 35/144 $ (TH chuột ở phòng 2 3 ngày liền thì xs qua phòng 1 là 1/3, TH không ở liền 3 ngày thì xs nhân 1/4)
* Ngày 4 phòng 2 có xác suất là: $1 - 13/64 - (35/144)* 2 = 179/576$ (Vì tổng xs các phòng trong 1 ngày là 1)

|  | **phòng 1** | **phòng 2** |**phòng 3**|**phòng 4**|
| --- | --- | --- | --- | --- |
| **Ngày 4** | 13/64 | 179/576 | 35/144 | 35/144 |
* Ngày 5 phòng 1 có xác suất là: $(13/64 - (1/2* 1/4* {(1/2)}^2))* 1/2 + (179/576 - 1/2* {(1/4)}^3)* 1/4 + (1/2* {(1/4)}^3)*1/3 = 757/4608 $ (Kế thừa phòng 1 ngày 4 không ở 3 ngày liền và 2 TH phòng 2 ngày 4 có hay không ở 3 ngày liền trước đó)

|  | **phòng 1** | **phòng 2** |**phòng 3**|**phòng 4**|
| --- | --- | --- | --- | --- |
| **Ngày 5** | 757/4608 |  |  |  |

**Xác suất con chuột vẫn ở Phòng 1 sau đó 5 ngày: 757/4608~ 0.1643**

In [11]:
# Mô phỏng

quantity = [0, 2, 4, 3, 3]
index = np.array([[0, 0, 0, 0, 0],
                [0, 1, 2, 0, 0],
                [0, 1, 2, 3, 4],
                [0, 2, 3, 4, 0],
                [0, 2, 3, 4, 0]] )

quantity2 = [0, 1, 3, 2, 2] # mảng dùng khi có phòng chuột đã ở 3 ngày
index2 = np.array([[0, 0, 0, 0, 0],
                [0, 2, 0, 0, 0],
                [0, 1, 3, 4, 0],
                [0, 2, 4, 0, 0],
                [0, 2, 3, 0, 0]] )# ma trận dùng khi có phòng chuột đã ở 3 ngày

sl = 0
n = 100000
for tt in range(n):
    start = 1 # Phòng chuột ở hiện tại
    frequence = 1 # Tầng suất chuột ở phòng hiện tại
    last = 1 # Phòng chuột ở trước đó 
    for _ in range(1, 6):
        if frequence < 3: # Khi bình thường
            percent = np.random.rand()
            for j in range(1, quantity[start] + 1):
                if (percent <= j*(1/quantity[start])):
                    start = index[start][j]
                    break
            if start == last:
                frequence += 1
            else:
                frequence = 1
                last = start

        elif frequence >= 3: # Tại phòng đó chuột đã ở 3 ngày
            percent = np.random.rand()
            for j in range(1, quantity2[start] + 1):
                if (percent <= j*(1/quantity2[start])):
                    start = index2[start][j]
                    break
            frequence = 1
            last = start

    if start == 1:
        sl += 1
        
print(sl/n)

0.16441


**d/ Tương tự Câu (c), làm lại Câu (b).**

In [12]:
# Mô phỏng
quantity = [0, 2, 4, 3, 3]
index = np.array([[0, 0, 0, 0, 0],
                [0, 1, 2, 0, 0],
                [0, 1, 2, 3, 4],
                [0, 2, 3, 4, 0],
                [0, 2, 3, 4, 0]] )
quantity2 = [0, 1, 3, 2, 2]
index2 = np.array([[0, 0, 0, 0, 0],
                [0, 2, 0, 0, 0],
                [0, 1, 3, 4, 0],
                [0, 2, 4, 0, 0],
                [0, 2, 3, 0, 0]] )

sl = np.array([0, 0, 0, 0, 0]) # Mảng để lưu số lượng chuột kết thúc ở phòng thứ i 
n = 5000
for _ in range(n):
    start = rd.randint(1, 4) # random phòng bắt đầu
    frequence = 1
    last = start 
    day = rd.randint(100, 200)
    for _ in range(day):
        percent = randpercent()
        if frequence < 3: # Khi bình thường
            for j in range(1, quantity[start]+1):
                if (percent <= j*(1/quantity[start])):
                    start = index[start][j]
                    break
            if start == last:
                frequence += 1
            else:
                frequence = 1
                last = start

        elif frequence >= 3:# Tại phòng đó chuột đã ở 3 ngày
            for j in range(1, quantity2[start]+1):
                if (percent <= j*(1/quantity2[start])):
                    start = index2[start][j]
                    break
            frequence = 1
            last = start

    sl[start] = plus_one(sl[start])

print(sl[1]/n, sl[2]/n, sl[3]/n, sl[4]/n)

0.154 0.3334 0.2608 0.2518


# Câu 3

**a/ Tính xác suất xe rời trạm 2 mà vẫn không có khách.**

In [13]:
# import thư viện cần thiết
import numpy as np
import random as rd

In [14]:
def randpercent():  # random phần trăm
    rand = rd.uniform(0, 1)
    return rand

**TH: Không có khách lên ở trạm 1 (Không có khách lên hay xuống ở trạm 2):**
* Xác suất: 0.5*0.5 = 0.25 *(1)*

**TH: Có 1 khách lên ở trạm 1 (Có 1 khách xuống và không có khách lên ở trạm 2):**
* Xác suất: 0.4*0.2 *0.5 = 0.04  *(2)*

**TH: Có 2 khách lên ở trạm 1 (Có 2 khách xuống và không có khách lên ở trạm 2):**
* Xác suất: 0.1*${0.2}^2$*0.5 = 0.002 *(3)*

**Xác suất xe rời trạm 2 không có khách:**
* (1) + (2) + (3) = 0.25 + 0.04 + 0.002 = 0.292

In [15]:
# Mô phỏng
up = np.array([0.1, 0.5, 1]) # Xác suất người lên xe 
quantity_up = np.array([2, 1, 0]) # số lượng người lên xe ứng với xác suất

def process(n, station):
    sl = 0
    for _ in range(n):
        passenger = 0 # số lượng người trên xe 
        for i in range(1,station+1):
            #get off the bus
            tmp_pas_bel = 0 # sl người xuống 
            for _ in range(passenger):
                percent = randpercent()
                if percent < 0.2:
                    tmp_pas_bel += 1
            passenger -= tmp_pas_bel

            # get on the bus
            percent = randpercent()
            for i in range(3): 
                if percent <= up[i]:
                    passenger += quantity_up[i]
                    break
        if passenger <= 0 : # Không có khách 
            sl += 1
    return sl

n = 100000
station = 2 # số lượng trạm dừng
print(process(n, station)/n)
# process(n, station)

0.29245


**b/ Tính xác suất xe rời trạm 3 mà vẫn không có khách.**

**TH: Xe rời trạm 2 không có khách (kế thừa câu a):**
* Xác suất: 0.292
* $\to$ TH này xe rời trạm 3 không có khách là: 0.292*0.5 = 0.146 *(1)*

**TH: Xe rời trạm 2 có 1 khách:**
* 0 khách lên trạm 1, 0 khách xuống trạm 2, 1 khách lên trạm 2 $\to$ có xác suất: 0.5*0.4 = 0.2
* 1 khách lên trạm 1, 0 khách xuống trạm 2, 0 khách lên trạm 2 $\to$ có xác suất: 0.4* 0.8 *0.5 = 0.16
* 1 khách lên trạm 1, 1 khách xuống trạm 2, 1 khách lên trạm 2 $\to$ có xác suất: 0.4* 0.2* 0.4 = 0.032 
* 2 khách lên trạm 1, 1 khách xuống trạm 2, 0 khách lên trạm 2 $\to$ có xác suất: (0.1* 0.2* 0.8* 0.5)* 2 = 0.016
* 2 khách lên trạm 1, 2 khách xuống trạm 2, 1 khách lên trạm 2 $\to$ có xác suất: 0.1* 0.2* 0.2* 0.4 = 0.0016
* $\to$ TH này xe rời trạm 3 không có khách là: (0.2 + 0.16 + 0.032 + 0.016 + 0.0016)* 0.2* 0.5 = 0.04096 *(2)*

**TH: Xe rời trạm 2 có 2 khách:**
* 0 khách lên trạm 1, 0 khách xuống trạm 2, 2 khách lên trạm 2 $\to$ có xác suất: 0.5*0.1 = 0.05
* 1 khách lên trạm 1, 1 khách xuống trạm 2, 2 khách lên trạm 2 $\to$ có xác suất: 0.4* 0.2* 0.1 = 0.008
* 1 khách lên trạm 1, 0 khách xuống trạm 2, 1 khách lên trạm 2 $\to$ có xác suất: 0.4* 0.8* 0.4 = 0.128
* 2 khách lên trạm 1, 0 khách xuống trạm 2, 0 khách lên trạm 2 $\to$ có xác suất: 0.1* 0.8* 0.8* 0.5 = 0.032
* 2 khách lên trạm 1, 1 khách xuống trạm 2, 1 khách lên trạm 2 $\to$ có xác suất: (0.1* 0.2* 0.8* 0.4)* 2 = 0.0128
* 2 khách lên trạm 1, 2 khách xuống trạm 2, 2 khách lên trạm 2 $\to$ có xác suất: 0.1* 0.2* 0.2* 0.1 = 0.0004
* $\to$ TH này xe rời trạm 3 không có khách là: (0.05 + 0.008 + 0.128 + 0.032 + 0.0128 + 0.0004)* 0.2* 0.2* 0.5 = 0.004624 *(3)*

**TH: Xe rời trạm 2 có 3 khách:**
* 1 khách lên trạm 1, 0 khách xuống trạm 2, 2 khách lên trạm 2 $\to$ có xác suất: 0.4* 0.8* 0.1 = 0.032
* 2 khách lên trạm 1, 0 khách xuống trạm 2, 1 khách lên trạm 2 $\to$ có xác suất: 0.1* 0.8* 0.8* 0.4 = 0.0256
* 2 khách lên trạm 1, 1 khách xuống trạm 2, 2 khách lên trạm 2 $\to$ có xác suất: (0.1* 0.2* 0.8* 0.1)* 2 = 0.0032
* $\to$ TH này xe rời trạm 3 không có khách là: (0.032 + 0.0256 + 0.0032)* 0.2* 0.2* 0.2* 0.5 = 0.0002432 *(4)*

**TH: Xe rời trạm 2 có 4 khách:**
* 2 khách lên trạm 1, 0 khách xuống trạm 2, 2 khách lên trạm 2 $\to$ có xác suất: 0.1* 0.8* 0.8* 0.1 = 0.0064
* $\to$ TH này xe rời trạm 3 không có khách là: 0.0064* 0.2* 0.2* 0.2* 0.2* 0.5 = 5.12.$10^{-6}$ *(5)*

**Xác suất xe rời trạm 3 mà vẫn không có khách:**
* (1) + (2) + (3) + (4) + (5) = 0.146 + 0.04096 + 0.004624 + 0.0002432 + 5.12.$10^{-6}$ = 0.19183232

In [16]:
# Mô phỏng
n = 100000
station = 3 # số lượng trạm dừng
print(process(n, station)/n)

0.19185


**c/ Tính xác suất xe rời trạm 5 với ít nhất 5 khách trên xe.**

In [17]:
# Mô phỏng
def process_c(n, station, guest):
    sl = 0
    for _ in range(n):
        passenger = 0 # số lượng người trên xe 
        for i in range(1,station+1):
            #get off the bus
            tmp_pas_bel = 0 # sl người xuống 
            for _ in range(passenger):
                percent = randpercent()
                if percent < 0.2:
                    tmp_pas_bel += 1
            passenger -= tmp_pas_bel

            # get on the bus
            percent = randpercent()
            for i in range(3): 
                if percent <= up[i]:
                    passenger += quantity_up[i]
                    break
        if passenger >= guest : # ít nhất lượng guest khách
            sl += 1
    return sl

n = 100000
station = 5 # số lượng trạm dừng
print(process_c(n, station, 5)/n)

0.03482


**d/ An xuống xe ở trạm 2, sau đó An thấy xe rời trạm 2 mà không có khách. Tính xác suất xe chỉ chở mình An khi xe rời trạm 1**

**Theo đề bài ta thấy rằng tại trạm 1 có 2 TH, 1 là mình An lên xe, 2 là An và 1 người khác sẽ lên.**
* TH 1: Mình An lên xe là p1.
* TH 2: An và 1 người khác sẽ lên là p2.
* Ta có p1 + p2 = 1, p1 = 4* p2
* Suy ra: p1 = 0.8, p2 = 0.2
* Vì dù thế nào khi rời trạm 2 sẽ không còn ai trên xe nên ta chỉ quan tâm ở trạm 1 xác suất mình An lên xe.
* Vậy xác suất xe chỉ chở mình An khi xe rời trạm 1 là p1 = 80%.


In [18]:
# Mô phỏng
n = 1000000
sl = 0
for i in range(n):
    percent = randpercent()
    if percent < 0.8:
        sl += 1
sl/n

0.800093

**e/ Tính kì vọng số khách trên xe khi xe rời trạm 2.**

In [19]:
def expectation(p): # tính kì vọng
    E = 0.0
    for i in range(len(p)):
        E += i*p[i]/n
    return E

**Gọi biến ngẫu nhiên X là số khách trên xe khi rời trạm 2. X nhận các giá trị là: 0, 1, 2, 3, 4 ứng với xác suất $p_0, p_1, p_2, p_3, p_4$.**

**Như đã tính ở câu b ta được:** 
* $p_0$ = 0.292
* $p_1$ = 0.2 + 0.16 + 0.032 + 0.016 + 0.0016 = 0.4096
* $p_2$ = 0.05 + 0.008 + 0.128 + 0.032 + 0.0128 + 0.0004 = 0.2312
* $p_3$ = 0.032 + 0.0256 + 0.0032 = 0.0608
* $p_4$ = 0.0064

**Kì vọng:**
* $E(X) = p_0*x_0 + p_1*x_1 + p_2*x_2 + p_3*x_3 + p_4*x_4$ = 0.292* 0 + 0.4096* 1 + 0.2312* 2 + 0.0608* 3 + 0.0064*4 = 1.08

In [20]:
# Mô phỏng
p = np.array([0, 0, 0, 0, 0]) # lưu xác suất từng giá trị Xi

def process_pro(n, station):
    for _ in range(n):
        passenger = 0
        for i in range(1,station+1):
            #get off the bus
            tmp_pas_bel = 0
            for _ in range(passenger):
                percent = randpercent()
                if percent < 0.2:
                    tmp_pas_bel += 1
            passenger -= tmp_pas_bel

            # get on the bus
            percent = randpercent()
            for i in range(3): 
                if percent <= up[i]:
                    passenger += quantity_up[i]
                    break
        p[passenger] += 1

    return p

n = 100000
sl = 0
station = 2
p = process_pro(n, station)
print(expectation(p))

1.0785200000000001


**f/ Tính kì vọng số khách trên xe khi xe rời trạm 3.**

**Gọi biến ngẫu nhiên X là số khách trên xe khi rời trạm 3. X nhận các giá trị là: 0, 1, 2, 3, 4, 5, 6 ứng với xác suất $p_0, p_1, p_2, p_3, p_4, p_5, p_6$.**

**Kế thừa ở câu b**

**Khi rời trạm 2 không có khách với xác suất 0.292** 
* Xác suất rời trạm 3 có 0 người: 0.292*0.5 = 0.146
* Xác suất rời trạm 3 có 1 người: 0.292*0.4 = 0.1168
* Xác suất rời trạm 3 có 2 người: 0.292*0.1 = 0.0292

**Khi rời trạm 2 có 1 khách với xác suất 0.4096**
* Xác suất rời trạm 3 có 0 người: 0.4016* 0.2* 0.5 = 0.04096
* Xác suất rời trạm 3 có 1 người: (0.4016* 0.8* 0.5) + (0.4016* 0.2* 0.4) = 0.196608
* Xác suất rời trạm 3 có 2 người: (0.4016* 0.2* 0.1) + (0.4016* 0.8* 0.4) = 0.139264
* Xác suất rời trạm 3 có 3 người: (0.4016* 0.8* 0.1) = 0.032768

**Khi rời trạm 2 có 2 khách với xác suất 0.2312**
* Xác suất rời trạm 3 có 0 người: 0.2312* 0.2* 0.2* 0.5 = 0.004624
* Xác suất rời trạm 3 có 1 người: (0.2312* 0.8* 0.2* 0.5)* 2 + (0.2312* 0.2* 0.2* 0.4) = 0.0406912
* Xác suất rời trạm 3 có 2 người: (0.2312* 0.8* 0.8* 0.5) + (0.2312* 0.8* 0.2* 0.4)* 2 + (0.2312* 0.2* 0.2* 0.1) = 0.1045024
* Xác suất rời trạm 3 có 3 người: (0.2312* 0.8* 0.8* 0.4) + (0.2312* 0.8* 0.2* 0.1)* 2 = 0.0665856
* Xác suất rời trạm 3 có 4 người: (0.2312* 0.8* 0.8* 0.1) = 0.0147968

**Khi rời trạm 2 có 3 khách với xác suất 0.0608**
* Xác suất rời trạm 3 có 0 người: (0.0608* 0.2* 0.2* 0.2* 0.5) = 0.0002432
* Xác suất rời trạm 3 có 1 người: (0.0608* 0.8* 0.2* 0.2* 0.5)* $C_{3}^1$ + (0.0608* 0.2* 0.2* 0.2* 0.4) = 0.00311296
* Xác suất rời trạm 3 có 2 người: (0.0608* 0.8* 0.8* 0.2* 0.5)* $C_{3}^1$ + (0.0608* 0.8* 0.2* 0.2* 0.4)* $C_{3}^1$ + (0.0608* 0.2* 0.2* 0.2* 0.1) = 0.01405696
* Xác suất rời trạm 3 có 3 người: (0.0608* 0.8* 0.8* 0.8* 0.5) + (0.0608* 0.8* 0.8* 0.2* 0.4)* $C_{3}^1$ + (0.0608* 0.8* 0.2* 0.2* 0.1)* $C_{3}^1$ = 0.02548736
* Xác suất rời trạm 3 có 4 người: (0.0608* 0.8* 0.8* 0.8* 0.4) + (0.0608* 0.8* 0.8* 0.2* 0.1)* $C_{3}^1$ = 0.01478656
* Xác suất rời trạm 3 có 5 người: (0.0608* 0.8* 0.8* 0.8* 0.1) = 0.00311296

**Khi rời trạm 2 có 4 khách với xác suất 0.0064**
* Xác suất rời trạm 3 có 0 người: (0.0064* 0.2* 0.2* 0.2* 0.2* 0.5) = 0.00000512
* Xác suất rời trạm 3 có 1 người: (0.0064* 0.8* 0.2* 0.2* 0.2* 0.5)* $C_{4}^1$ + (0.0064* 0.2* 0.2* 0.2* 0.2* 0.4) = 0.000086016
* Xác suất rời trạm 3 có 2 người: (0.0064* 0.8* 0.8* 0.2* 0.2* 0.5)* $C_{4}^2$ + (0.0064* 0.8* 0.2* 0.2* 0.2* 0.4)* $C_{4}^1$ + (0.0064* 0.2* 0.2* 0.2* 0.2* 0.1) = 0.00055808
* Xác suất rời trạm 3 có 3 người: (0.0064* 0.8* 0.8* 0.8* 0.2* 0.5)* $C_{4}^1$ + (0.0064* 0.8* 0.8* 0.2* 0.2* 0.4)* $C_{4}^2$ + (0.0064* 0.8* 0.2* 0.2* 0.2* 0.1)* $C_{4}^1$ = 0.00172032
* Xác suất rời trạm 3 có 4 người: (0.0064* 0.8* 0.8* 0.8* 0.8* 0.5) + (0.0064* 0.8* 0.8* 0.8* 0.2* 0.4)* $C_{4}^1$ + (0.0064* 0.8* 0.8* 0.2* 0.2* 0.1)* $C_{4}^2$ = 0.0024576
* Xác suất rời trạm 3 có 5 người: (0.0064* 0.8* 0.8* 0.8* 0.8* 0.4) + (0.0064* 0.8* 0.8* 0.8* 0.2* 0.1)* $C_{4}^1$ = 0.00131072
* Xác suất rời trạm 3 có 6 người: (0.0064* 0.8* 0.8* 0.8* 0.8* 0.1) = 0.000262144

**Ta được xác suất:**
* $p_0$ = 0.146 + 0.04096 + 0.004624 + 0.0002432 + 0.00000512 = 0.19183232
* $p_1$ = 0.1168 + 0.196608 + 0.0406912 + 0.00311296 + 0.000086016 = 0.357298176
* $p_2$ = 0.0292 + 0.139264 + 0.1045024 + 0.01405696 + 0.00055808 = 0.28758144
* $p_3$ = 0.032768 + 0.0665856 + 0.02548736 + 0.00172032 = 0.12656128
* $p_4$ = 0.0147968 + 0.01478656 + 0.0024576 = 0.03204096
* $p_5$ = 0.00311296 + 0.00131072 = 0.00442368
* $p_6$ = 0.000262144 = 0.000262144

**Kì vọng:**
* $E(X) = p_0*x_0 + p_1*x_1 + p_2*x_2 + p_3*x_3 + p_4*x_4$ = 0.19183232* 0 + 0.357298176* 1 + 0.28758144* 2 +0.12656128* 3 + 0.03204096* 4 + 0.00442368* 5 + 0.000262144*6 

= 1.464

In [21]:
p = np.array([0, 0, 0, 0, 0, 0, 0]) # lưu xác suất từng giá trị Xi

n = 100000
sl = 0
station = 3

p = process_pro(n, station)
print(expectation(p))

1.46238


**g/ Giả sử xe chứa tối đa 5 khách, nếu lượng khách lên nhiều thì xe chỉ cho phép đủ người lên. Tính xác suất xe rời trạm 5 với đủ khách.**

**Tương tự như câu b và câu f ta thấy được:**
* Tại trạm n có 0 khách kế thừa trạm n-1 cố 0 -> 5 khách
* Tại trạm n có 1 khách kế thừa trạm n-1 cố 0 -> 5 khách
* Tại trạm n có 2 khách kế thừa trạm n-1 cố 0 -> 5 khách
* Tại trạm n có 3 khách kế thừa trạm n-1 cố 1 -> 5 khách
* Tại trạm n có 4 khách kế thừa trạm n-1 cố 2 -> 5 khách
* Tại trạm n có 5 khách kế thừa trạm n-1 cố 3 -> 5 khách

**Ta lập được bảng xác suất sau:**
|  | **Trạm 1** | **Trạm 2** | **Trạm 3** | **Trạm 4** | **Trạm 5** |
| --- | --- | --- | --- | --- | --- |
| 0 Khách | 0.5 | 0.292 | 0.1918 | 0.1376 | 0.1064 |
| 1 Khách | 0.4 | 0.4096 | 0.3573 | 0.3057 | 0.2651 |
| 2 Khách | 0.1 | 0.2312 | 0.2876 | 0.3027 | 0.3006 |
| 3 Khách | 0 | 0.0608 | 0.1266 | 0.1737 | 0.2033 |
| 4 Khách | 0 | 0.0064 | 0.0320 | 0.0629 | 0.0908 |
| 5 Khách | 0 | 0 | 0.0044 | 0.0169 | 0.0337 |

**Vậy xác suất xe rời trạm 5 với đủ khách: ô (5 khách, trạm 5) = 0.0337**

In [22]:
# Mô phỏng
p = np.array([0, 0, 0, 0, 0, 0])

def process_g(n, station):
    for _ in range(n):
        passenger = 0
        for i in range(1,station+1):
            #get off the bus
            tmp_pas_bel = 0
            for _ in range(passenger):
                percent = randpercent()
                if percent < 0.2:
                    tmp_pas_bel += 1
            passenger -= tmp_pas_bel

            # get on the bus
            percent = randpercent()
            for i in range(3): 
                if percent <= up[i]:
                    passenger += quantity_up[i]
                    if passenger > 5:
                        passenger = 5
                    break
        p[passenger] += 1
    return p

n = 100000
station = 5
p = process_g(n, station)
print(p[5]/n)

0.03387


**h/ Tương tự Câu (g), tính kì vọng số khách trên xe khi xe rời trạm 5.**

**Gọi biến ngẫu nhiên X là số khách trên xe khi rời trạm 5. X nhận các giá trị là: 0, 1, 2, 3, 4, 5 ứng với xác suất $p_0, p_1, p_2, p_3, p_4, p_5$.**

**Như đã tính ở câu g ta được:** 
* $p_0$ = 0.1064
* $p_1$ = 0.2651
* $p_2$ = 0.3006
* $p_3$ = 0.2033
* $p_4$ = 0.0908
* $p_5$ = 0.0337

**Kì vọng:**
* $E(X) = p_0*x_0 + p_1*x_1 + p_2*x_2 + p_3*x_3 + p_4*x_4 + p_5*x_5$ = 0.1064* 0 + 0.2651* 1 + 0.3006* 2 + 0.2033* 3 + 0.0908* 4 + 0.0337* 5 = 2.0079

In [23]:
# Mô phỏng
p = np.array([0, 0, 0, 0, 0, 0])

n = 100000
station = 5
p = process_g(n, station)
expectation(p)

2.01195